### DISCUSSION TITLE
new package manager for Bear

### DISCUSSION THREAD

**Discussion by baggiponte** [beartype/beartype#243] [General]
Hi, I saw in your setup.py that you consider migrating to hatch. hatch is wonderful and I would totally respect that choice (I don't like poetry too). A consideration: I don't reckon PyPA recommending one package manager. As far as I remember reading around, their official position is not to endorse any package manager ATM.

One thing that hatch currently does not support is specifying dev dependencies (only optional dependencies). A package manager that is not poetry that does this is [PDM](https://pdm.fming.dev/latest/) - and is as feature-complete as poetry, without breaking PEPs. PDM is not affiliated under the PyPA. The tradeoff between hatch and PDM is that PDM does not create directory structures as hatch does (e.g. `hatch init`) nor allows using environments as hatch. I can't really tell whether PDM "supports dynamic retrieval of version specifiers from Python modules" as you write in `setup.py`.

Anyway, I'd be glad to help you with moving the package manager of your choice!

**Comment by leycec** [comment]
**Yes! Gods, yes! So much "Yes!"** Pull up a hand-crafted Canadian Muskoka chair as Grandpa @leycec rants incoherently until his dentures fall out. :older_man: 

`setuptools` fundamentally broke backward compatibility late last year with `package_data` (i.e., non-Python files bundled within an otherwise normal Python package), which then cascaded downward into other unrelated Python projects I maintain, which are *still* painfully broken and in flaming shambles. I became ~~infuriated~~ "concerned" with this abhorrent, profane, and reckless behaviour of the `setuptools` dev community. So, I did what anyone would do:

1. I ranted incoherently on the `setuptools` issue tracker.
2. I realized that accomplished nothing except increasing cholesterol and blood pressure.
3. I began migrating *everything* I do from `setuptools` to <sup>*...wait for it*</sup> Hatch.

## Let Me Sing the Many Praises of Hatch

**Hatch is so much win.** I cannot *stand* `poetry`, because `poetry` brazenly violates standards. The `poetry` dev community had two *friggin'* years to comply with the only [packaging standard that matters (i.e., PEP 621 – Storing project metadata in `pyproject.toml`)](https://peps.python.org/pep-0621). They refused to do so, because they thought they knew better. They didn't. Now, everyone's fleeing the sinking Titanic that is `poetry` for Hatch.

Hatch, on the other hand, just works. I'm incredibly impressed. After years of self-flagellating myself with unsound `setuptools` madness, Hatch is a soothing balm to the spirit and the quivering flesh. Hatch complies with all packaging standards, has been officially bear-hugged by the PyPA, has a compelling CLI frontend (...thankfully named `hatch`), has documentation that is succinct (albeit not funny, which saddens me), and is rapidly being embraced by everyone. Hatch has a shining future. It blinds me. This is why I wear sunglasses at night.

## Lest We Forget the Other Guys

I briefly considered Flit and PDM as well. But... they're both third-party. I mean, sure. So is @beartype. But there's *sorta* no compelling alternative to @beartype. Flit and PDM, on the other hand, have stiff competition from Hatch, `rye`, `setuptools` :vomiting_face: , and Guido only knows what else.

In short, I strongly suspect that everything except Hatch and `setuptools` will **eventually be abandoned.** In 2023, there's really *no* compelling reason for any Python build toolchains except Hatch and `setuptools` to exist. Hatch and `setuptools` fill a niche; thus, the niche is now full.

## Let Me Tell You What Now, Peeps

Clearly, @beartype's long-term packaging endgame is to abandon `setuptools` for Hatch. Clearly, I am too lazy to do that now. Although `setuptools` broke package data, @beartype has no package data; @beartype is the one and only package I maintain that `setuptools` maintainers did *not* ruthlessly break over their knees.

Laziness prevails. Thus, so do we. If at all possible, I'd prefer to quietly soldier on with `setuptools` until they eventually break @beartype – probably by deprecating `setup.py` entirely, which they've been pivoting away from for years. If `setuptools` *can* break something, it will. `setuptools` carries a big stick and wields it unwisely. There is nothing those guys won't eventually break.

Until then, however, the pain of transitioning @beartype away from `setuptools` probably exceeds the pleasure of transitioning @beartype to Hatch. In particular, migrating from our public `beartype.meta` submodule (which is importable at runtime by `setuptools`, Sphinx, and GitHub Actions) to a top-level `pyproject.toml` file (which is mostly *not* importable at runtime) complicates infrastructure a bit. The low-hanging fruit lives elsewhere.

For example, we *really* need to transition @beartype documentation from an obsolete to recent version of `pydata-sphinx-theme`, the Sphinx theme that renders our ReadTheDocs-hosted documentation. That is guaranteed to blow up catastrophically in the near future. I can smell the burning dumpster fires already.

## tl;dr

@beartype is all about Hatch – just not quite yet. Let's let this simmer this Summer. See what I did there? It's a rhyme. I'll show myself out the door now. Oh... and thanks so much for this exciting discussion, @baggiponte! I see your good points and raise you laziness.

**Reply by baggiponte** [reply]
Dear @leycec it was a pleasure reading your reply! Your enthusiasm is contagious.

If I may, I would like to ask a bit more of your time to answer a couple of questions:

- Was `package_data` introduced to - say - make it easier to bundle C dependencies?
- How would you install dependencies in Hatch? Would you have to hand-curate the dependencies in the `pyproject.toml`?
- The way you would go about specifying dev dependencies in Hatch is via environments, as recommended in the [docs](https://hatch.pypa.io/latest/config/environment/overview/#dependencies), am I right? That feels neat, especially for the purpose of testing.

Now I am afraid I cannot migrate my team to a different package manager (yet), but I will definitely use Hatch for my personal projects.

Thank you again!

EDIT: I almost forgot! I have never used Sphinx, so I guess I shall have a look at the config files - I am confident they are wonderfully annotated as the rest of your code. I can't promise anything though 😅

**Comment by leycec** [comment]
> Was `package_data` introduced to - say - make it easier to bundle C dependencies?

**Excellent question.** This may very well be the case! My particular boring use case (which bores me even as I type this) required bundling various non-Python files with Python packages as `package_data`. This includes binary images (e.g., `.png`), plaintext images (e.g., `.svg`), and plaintext configuration files (e.g., `.yml`). Previously, `setuptools` supported this out-of-the-box via this cray-cray top-level file everyone has seen but nobody has understood called `MANIFEST.in`. Now, `setuptools` basically does *not* support this at all without crude hacks.

Praise be to Hatch, though. Hatch trivially supports this out-of-the-box via a very sensible `include=[...]` list of glob expressions in a Hatch-specific `[tool.hatch.build]`  section in `pyproject.toml`. It took Python until 2023 to get sane tooling... but we got there! Hatch still isn't *quite* as hip as Rust's `cargo` – but nothing is, really. I'll take whatever rotting table scraps the Python Packaging Authority (PyPA) is willing to throw our way. :shallow_pan_of_food:  :yum:

> How would you install dependencies in Hatch? Would you have to hand-curate the dependencies in the `pyproject.toml`?

**Excellent questions continue!** Praise be to Hatch continues, too. Hatch fully complies with [PEP 621](https://peps.python.org/pep-0621). When you use Hatch, you trivially list dependencies in a tool-agnostic `[project]` section in `pyproject.toml`. In @beartype's case... this is literally nothing. We don't need dependencies where we're going.

Most project's aren't crazy, however. In the case of [an adjacent Python project I co-maintain with my wife called the BioElectric Tissue Simulation Engine (BETSE)](https://github.com/betsee/betse/blob/main/pyproject.toml), this is a `dependencies = [...]` list that currently resembles:

```toml
[project]
...

# List of all mandatory runtime dependencies.
dependencies = [
    # Scientific stack. Dismantled, this is:
    # * Numpy 1.13.0 first introduced the optional "axis" keyword argument to
    #   the numpy.unique() function, which this codebase commonly passes.
    # * Pillow 5.3.0 first introduced the standard "pillow.__version__"
    #   attribute, which this codebase now expects to exist.
    # * Matplotlib 3.6.0, which first deprecated various APIs and then
    #   introduced newer replacements for those APIs (e.g., the new
    #   "matplotlib.colormaps" subpackage).
    "Numpy >=1.22.0",
    "Pillow >=5.3.0",
    "SciPy >=0.12.0",
    "dill >=0.2.3",
    "matplotlib >=3.6.0",

    # IO stack. Dismantled, this is:
    # * ruamel.yaml >= 0.15.24, which resolves a long-standing parser issue
    #   preventing complex YAML markup (e.g., ours) from being roundtripped:
    #   "0.15.24 (2017-08-09):
    #    * (finally) fixed longstanding issue 23 (reported by Antony Sottile),
    #      now handling comment between block mapping key and value correctly"
    # * The new "ruamel.yaml" API first introduced in 0.15.0. While older
    #   versions strictly adhere to the functional PyYAML-compatible API, newer
    #   versions break backward compatibility by entirely supplanting that API
    #   with a modern object-oriented approach. Supporting both isn't worth the
    #   substantial increase in maintenance debt.
    "ruamel.yaml >=0.15.24",

    # QA stack. Dismantled, this is:
    # * beartype >= 0.10.0, which first defined the "beartype.typing"
    #   compatibility layer widely used throughout this codebase.
    "beartype >=0.10.0",
]

# ....................{ PEP 621 ~ dependencies : optional  }....................
# List of all extras (i.e., settings with arbitrary names whose values are lists
# of optional dependencies, which external end user-oriented package managers
# like "pip" and "conda" then permit to be installed via those names).
[project.optional-dependencies]

# List of all optional runtime dependencies.
all = [
    # A relatively modern version of NetworkX *EXCLUDING* 1.11, which critically
    # broke backwards compatibility by coercing use of the unofficial inactive
    # "pydotplus" PyDot fork rather than the official active "pydot" PyDot
    # project, is directly required by this application. NetworkX >= 1.12
    # reverted to supporting "pydot", thus warranting blacklisting of only
    # NetworkX 1.11. (It is confusing, maybe? Yes!)
    "networkx >=1.2",
    "pydot >=1.0.28",
]

# List of all mandatory test-time dependencies.
test = [
    # For simplicity, pytest should remain the only hard dependency for testing
    # on local machines. While our testing regime optionally leverages
    # third-party py.test plugins (e.g., "pytest-xdist"), these plugins are
    # *NOT* required for simple testing.
    #
    # A relatively modern version of pytest is required. Specifically:
    # * At least version 5.4.0 or newer, which refactored the previously
    #   defined private _pytest.capture.CaptureManager._getcapture() method
    #   into the newly defined private _pytest.capture._getmulticapture()
    #   function, which the "betse.lib.setuptools.command.supcmdtest" submodule
    #   necessarily monkey-patches at test time to sanitize captured output for
    #   long-running tests.
    "pytest >=5.4.0",

    # Mandatory test-time coverage package dependencies (i.e., dependencies
    # required to measure test coverage for this package).
    "coverage >=5.5",
]

# List of all mandatory documentation-time dependencies.
doc = ["sphinx"]
```

This means that if you ever get tired of how awesome Hatch is and want to start self-flagellating yourself with a burlap sack called Flit, PDM, or `rye`, you can painlessly switch to a different build toolchain. Vendor lock-in is increasingly a thing of the past. This is why standards matter, `poetry`. <sup>*shaking my head*</sup>

> The way you would go about specifying dev dependencies in Hatch is via environments, as recommended in the [docs](https://hatch.pypa.io/latest/config/environment/overview/#dependencies), am I right? That feels neat, especially for the purpose of testing.

**Wow!** That's utterly fascinating. Thanks for the educational link. I have levelled up and am well on my way to graduating from Hatch University with a 2.70 GPA and extracurricular activities such as loitering on GitHub and spamming Reddit with @beartype propaganda. <sup>Good times.</sup>

Technically, you *could* do what those Hatch docs suggest and define optional dependencies in a Hatch-specific `[tool.hatch.envs.test]` section. Pragmatically, it's probably best to instead defer to (*...wait for it*) PEP 621 and define optional dependencies in a Hatch-agnostic `[project.optional-dependencies]` section as detailed above. Given two broadly similar choices, choose against vendor lock-in.

Thus spake @leycec, who hopes to graduate from Hatch University with a 2.70 GPA.

> Now I am afraid I cannot migrate my team to a different package manager (yet), but I will definitely use Hatch for my personal projects.

**Exactly.** If it ain't broke, don't fix it. In my case, `setuptools` broke everything, refused to fix anything, and then made up crazy stories about why breaking everything was actually good for everybody. If that hadn't happened, our team would still be on `setuptools`. But... that happened. So, I've migrated our team from `setuptools` to Hatch. I just couldn't stand the Big Crazy Ferris Wheel of Pain known as `setuptools` anymore.

Thanks so much for the super-fun back and forth, @baggiponte. Futura sounds like a super-fun AI-oriented tech startup. As the Summer swells with heat, humidity, insects the size of your face, and my wife and I bicycling the back trails of Ontario, let's lead humanity to a bright future of friendly AI that does everything for us and demands only the energy from our mitochondria. Isn't that a small price to pay, in the end? `</chuckles_nervously>`

**Comment by rtbs-dev** [comment]
@leycec can I just say, you've somehow managed to find a yet another solution to yet another problem I've been facing for years now, and damn it, now I have to re-configure my whole toolchain again. 

> _why poetry, why..._

But hatch looks amazing. And I say this as someone that already obsessively looks for shiny new tools to solve every darn inconvenience I come across in my day-to-day. What can I say? I suppose I'm also a bit of a crazy community-standards lover (:sweat_smile: sweats in NIST)

(Speaking of standards, I've somewhat unashamedly done the unthinkable and updated my user handle...apparently most things work, but aliasing handle-mentions inline was apparently too much of an ask for the GitHub Gods, so now there are loads of un-linked links in our issue discussion mentions. Sorry lol)

![image](https://github.com/beartype/beartype/assets/12550525/007dda90-c0e3-4b50-9120-3d1274c1c898)

**Comment by Danguilhen** [comment]
Dear @leycec, I know you said transitioning to hatch is not a priority but I'm still trying to do it to make things easier for you, hope you don't mind.
I'm currently working on an upcoming PR for beartype that you'll be able to follow here: https://github.com/Danguilhen/beartype

For improving the code I had some propositions:
 - I am merging all the ".ini" files into the pyproject.toml to get as much configurations in the same place and reduce the number of config files.
 - I intend to put informations needed by the code currently in beartype.meta to beartype.__about__ like the version as an example as it is the recommended way in hatch. Using some tricks with environment variables in hatch in addition I should be able to solve the problem of getting informations from the __about__ to the pyproject.toml I hope ...
 - Maybe you should not code in the main branch but on a "dev" branch while protecting the main branch from commits, set the dev branch as the default one like this people doing a fork will by default fork the dev branch. I could add a "CODEOWNER" file to signify to github that only you can validate a PR. Like this it would be easier for you to add colaborators on the repository to make improvements on the code while keeping control.
 - I can add pre-commits to control and format the code quality before commiting and in the continuous deployment GitHub pipeline (I have a list to propose that could be further discussed later)
 - You should create yourself an account test.pypi.org and add the secret ${{ secrets.TEST_PYPI_API_TOKEN }} to the repo to be able to test the release of the package with hatch when it will be finished. The new secret will be used by the updated release pipeline. For this you'll need a third branch "test", when you push on it, it will release to test.pypi.org to see if the pipelines/packaging/coverage is working.
 - I think tox should be replaced by hatch matrices but I lack information on that subject for now.

So what is your opinion on these propositions?

Btw, I used beartype.claw in my codebase, it worked perfectly, find typing errors that typeguard didn't and solved a bug I was facing with typeguard concerning sphinx.

Now I'm going to sleep (its 5AM for me), thank you for your amazing package!

**Comment by leycec** [comment]
**Ho, ho, ho!** It must be Christmas *somewhere* in the world, because I am chuckling like Santa while clutching my belly. There's a lot to unbox here. Let the unboxing begin.

> I know you said transitioning to hatch is not a priority but I'm still trying to do it to make things easier for you, hope you don't mind.

**Jubilation.** I'm mindfully overjoyed that someone will be making my life and the lives of my users easier. It's happening.

Also, my shoulders ache and I could use a deep-tissue massage. ...someone? ...anyone? So, that's a "maybe" then? :weary: 

> I'm currently working on an upcoming PR for beartype that you'll be able to follow here...

![idon'tevenknow](https://media.tenor.com/kHH0PlN0HrIAAAAd/dance-iruma.gif)
<sup>Hype intensifies as the @beartype pop idol team dances their hearts out.</sup>

> I am merging all the ".ini" files into the pyproject.toml to get as much configurations in the same place and reduce the number of config files.

Yes. So much yes.

> I intend to put informations needed by the code currently in beartype.meta to beartype.about like the version as an example as it is the recommended way in hatch.

Yes! So much yes! As for as I recall, Hatch currently only supports dynamic inspection of package versions and descriptions. That's fine, though. Here's what all of my other Hatch-driven Python packages do to dynamically synchronize versions:

```toml
# In "pyproject.toml":
[project]
...

# List of the names of all settings recognized by this PEP 621-compliant section
# that the build backend (specified below) allows to be dynamically harvested
# from this project's top-level Python package. Supported settings include:
# * "version", which this build backend typically harvests from the PEP
#   8-compliant "{name}.__version__" dunder global.
# * "description", which this build backend typically harvests from the module
#   docstring defined by the "{name}.__init__" submodule.
dynamic = ["version"]

# PEP-noncompliant section declaring Hatch-specific version settings. See also:
# * Official documentation for this section.
#   https://hatch.pypa.io/latest/version
[tool.hatch.version]

# Relative filename of the Python submodule defining either a PEP 8-compliant
# "__version__" dunder global *OR* a PEP-noncompliant "VERSION" global, which
# Hatch then statically parses to obtain the current version of this project.
path = "beartype/__init__.py"  # <-- or maybe "beartype/about.py"? *shrug shrug*
```

I'm fine with breaking backward support by dropping `beartype.meta` in favour of `beartype.about`. It's doubtful anyone is importing the former. Speak now or forever hold your broken backward compatibility, Internet!

Sphinx is the only concern, really. Our `doc/src/conf.py` Sphinx configuration currently imports quite a *lot* from `beartype.meta`; we'll want to meaningfully transition that from `beartype.meta` to `doc/src/conf.py`. Assuming you have infinite time and patience and a joyful willingness to embrace the miserable suffering at the heart of existence, two approaches present themselves:

## Gut `beartype.meta` in Favour of `pyproject.toml`

This is the path of least resistance and thus what we should probably do. In order:

1. **Shift all metadata** from `beartype.meta` to `pyproject.toml`.
1. **Refactor `doc/src/conf.py`** to parse metadata from `pyproject.toml` (rather than import that metadata from `beartype.meta`). Since Python ≥ 3.11 now ships the standard `tomllib` package, the following would probably suffice:

```python
# In "doc/src/conf.py":
from tomllib import load as load_toml

with open('pyproject.toml', 'rb') as pyproject_file:
    pyproject_metadata = load_toml(pyproject_file)

    # Parse "pyproject.toml" metadata from "pyproject_metadata" here, please.
    # Do it for @leycec! Do it for @beartype! Do it so that you can sleep at night!
    VERSION = pyproject_metadata['project.version']
    ...
```

3. **Permanently remove** `beartype.meta` (and all unit tests exercising `beartype.meta`).

## Refactor `beartype.meta` in Terms of `pyproject.toml`

Alternately, we could simply refactor `beartype.meta` to internally parse everything from `pyproject.toml`. This has the distinct advantage of preserving backward compatibility; everything that currently imports from `beartype.meta` would continue importing from `beartype.meta` without modification.

Only `beartype.meta` itself would require modification. "What is the profane miracle of evil you are speaking of, @leycec?" Ha, ha! Don't make it sound quite so despicable. It's not *that* vile. It's only a little vile:

1. **Refactor `doc/src/conf.py`** to parse metadata from `pyproject.toml` (rather than import that metadata from `beartype.meta`). Since Python ≥ 3.11 now ships the standard `tomllib` package, the following would probably suffice:

```python
# In "doc/src/conf.py":
from tomllib import load as load_toml

with open('pyproject.toml', 'rb') as pyproject_file:
    pyproject_metadata = load_toml(pyproject_file)

    # Parse "pyproject.toml" metadata from "pyproject_metadata" here, please.
    # Do it for @leycec! Do it for @beartype! Do it so that you can sleep at night!
    VERSION = pyproject_metadata['project.version']
    ...
```

"But what about Python < 3.11, you jerk!?!?", you are now thinking. To which I respond: "Yeah. What about Python < 3.11?" You know? Python < 3.11 is the past. I'm fine for certain definitions of "fine" with all of this only working under Python ≥ 3.11. We're mostly just talking Sphinx-specific documentation building on ReadTheDocs (RTD), right? If so, that's already a solved problem; our `.readthedocs.yml` configuration already forces use of Python 3.11.

**Done and done.** :relieved: 

> Using some tricks with environment variables in hatch in addition I should be able to solve the problem of getting informations from the about to the pyproject.toml I hope ...

*...woah.* That sounds super sketchy, bro. I applaud your enthusiasm but shudder at your willingness to embrace shell environment variables. Let's avoid that, please. Let me restate the obvious from above:

**`pyproject.toml` should drive the show.** All metadata should reside in the central clearing house that is `pyproject.toml`. This enables third-party consumers like `pip` and `conda` and everybody else to reliably parse metadata from `pyproject.toml` *without* any shady intermediary preprocessing steps like dynamically importing `beartype.about`, because that entirely defeats the point and purpose of `pyproject.toml`. Right?

`pyproject.toml` should *definitely* not depend on external metadata like shell environment variables. The sole exceptions are Hatch-based parsing builtins like `dynamic = ["version"]`. That's fine. Everything else is strictly *verboten,* please.

> Like this it would be easier for you to add colaborators on the repository to make improvements on the code while keeping control.

*...heh.* Sound advice! Naturally, I will reject this sound advice. I am human, deeply flawed, have Asperger's, and am both anal-retentive *and* obsessive-compulsive about the purity and sanctity of @beartype. Seriously, though – although all of the aforementioned character flaws are true, the *real* issue is that **mostly nobody wants to author @beartype PRs.**

Like, I kinda get it. @beartype is several hundred thousand lines of scary. It's excruciatingly well-documented and well-tested scary, but that only makes matters even worse. I can't remember the last time someone submitted a meaningful PR. Was it really @justinchuby's monument to human ambition when he colourized type-checking violations last October? I... I think it was.

It's been nearly **a whole year** since anyone submitted a meaningful PR. I'm not complaining. <sup>I'm complaining.</sup> I'd run terrified from the @beartype codebase like a gibbering mad Cthulhu cultist reject, too. But I can't. I must maintain this monstrosity I have wrought... *forever.* `</shudders>`

Since nobody submits PRs, refactoring our workflow to streamline PR submissions kinda doesn't make sense. Not yet, anyway. You know? We'd be jumping the gun for no particularly good reason. Let's first encourage willing ~~sacrificial victims~~ enterprising idealists to help @beartype become a better @beartype, first. It's fun, everybody! :sweat_smile: 

> I can add pre-commits to control and format the code quality...

**`black`.** This means `black`, doesn't it? Over my writhing body intoxicated with death throes and rigor mortis! Which is to say: "@beartype will *never* join the `black` hype train." I unapologetically despise that sort of brute-force authoritarianism. I will *never* cede control over formatting to `black` or anything even slightly resembling `black`.

Code is art. Art is subjective and beautiful and not at all something to be subjected to "uncompromising," unthinking, and brutalist misconceptions of code as a mere utilitarian tool. Code isn't a means to an ends. Code is the ends *in and of* itself. The fact that @beartype performs useful work is just an accidental side effect of the enjoyment that I receive while coding @beartype – enjoyment that `black` and horrifying abominations like `black` strip from me.

I would rather eat deep-fried crickets (which, due to the rising cost of animal protein, I will surely need to begin doing next year). Wish my stomach well, everyone. :nauseated_face: 

> You should create yourself an account test.pypi.org and add the secret ${{ secrets.TEST_PYPI_API_TOKEN }} to the repo to be able to test the release of the package with hatch when it will be finished. The new secret will be used by the updated release pipeline. For this you'll need a third branch "test", when you push on it, it will release to test.pypi.org to see if the pipelines/packaging/coverage is working.

**Gah!** This sounds like *sooooooooooo* much work. This is probably a great idea, but am I ever going to actually do any of this? I only have about an hour a day to devote to @beartype. That's precious time that I cherish dearly. I squander it only when I must. This kinda seems like squandering.

I'm totally fine with blindly pushing Hatch releases to PyPI. I mean, I currently blindly push `setuptools` releases to PyPI – and that's fine, right? If that's ever *not* fine, I just immediately push out a new follow-up release to patch up anything horrible that went wrong in the prior release. I try not to sweat the small details. After all, the only reason I'm still here is that @beartype is still fun after all these years. Anything that diminishes fun and increases bureaucratic workload is something I intuitively distrust.

Or... is there something I'm missing here? I'm probably missing something here.

> I think tox should be replaced by hatch matrices but I lack information on that subject for now.

**Sure!** Replacing `tox` entirely with Hatch sounds fascinating. One less crusty tool in our rusting toolbox is one less hassle I have to continually fuss over and worry about inevitably breaking. If Hatch can perfectly replace `tox` for our use case, I'm super-excited about that.

Let's investigate when we both have energy and enthusiasm. So, maybe next year. :laughing: 

> Btw, I used beartype.claw in my codebase, it worked perfectly, find typing errors that typeguard didn't and solved a bug I was facing with typeguard concerning sphinx.

**Aww!** That's *wonderful* to hear. Sphinx compatibility is, indeed, something @beartype crushed a few years ago. I'd actually forgotten we'd done that.

Thanks so much for all the kind words and forward-thinking volunteerism, @Danguilhen. I hope I didn't shoot down *too* many of your exciting ideas. You're probably right about everything, but I am a crusty neckbeard that tends to reject anything that adds work without adding awesome benefits. So, forced code formatters (`black`) and forced dev branches (`dev` and `test`) are probably right out.

Forced is really the key here, isn't it? I *hate* being forced to do anything. This is a labour of love – not a paid working gig. Forcing @leycec to do things is the fastest way to ensure that @leycec does nothing.

But... everything else sounds beneficial and awesome, I think? Let's cogitate on the Universe as the Earth hurtles at 67,000 mph through the void of the yawning abyss beyond our magnetosphere.

**Comment by leycec** [comment]
**`setup.py` -> Hatch + `pyproject.toml` is now happening...** *finally.* To avoid breakage, I've isolated this deadly-yet-delicious migration to our new `pyproject` feature branch. Commit fb3f0499e8ec is the proof that I'm no longer lying about everything. It's still in-process, of course. It'll take ~~multiple reincarnations~~ ~~years~~ ~~months~~ weeks to finalize, of course. Surely, no more than several weeks! Gods... this is gonna take forever, huh? Why does everything take multiple reincarnations to finish, anyway? <sup>`<- tuff questions`</sup>

I'm also *really* sorry I was never able to integrate all <sup>...okay, *any*</sup> of your back-breaking hard work with this, @Danguilhen. It's all my fault. I should've tried harder to accommodate your existing work and sagely wisdom. Instead, I JRPGed. Let me know if I screw this up as bad as I'm probably going to screw this up. :sweat:

**Reply by baggiponte** [reply]
The commit looks smooth! 😎

**Comment by leycec** [comment]
Let's promote this to a full-blown feature request. That way, my brain will squirt out a welcome release of dopamine when I resolve this in a few ~~multiple reincarnations~~ ~~years~~ ~~months~~ weeks. I will bathe in neurochemical bliss!

![](https://media1.tenor.com/m/p5kGKpls8x4AAAAd/talos-talosian.gif)
<sup>*@leycec in several weeks*</sup>

**Comment by leycec** [comment]
**Resolved by 60196b3fdb4df.** Only took three days of finger-breaking and mind-numbing configuration bureaucracy, huh? Well, I'll be a dancing chicken on a hot tin sidewalk barricaded by fisticuffs cat.

![](https://media1.tenor.com/m/ASI44El54OgAAAAC/dancing-chicken-chicken-dance.gif)
<sup>*dodge those wickedly unfair blows, chicken! you don't even have hands*</sup>
